### ISSUE TITLE
isl conflicts with defaults gcc

### ISSUE THREAD

**Issue by jakevdp** [conda-forge/isl-feedstock#3]
Causes failures in gcc unless the version is pinned to use the default channel instead. See details here: https://github.com/conda-forge/megaman-feedstock/pull/3

**Comment by frol** [comment]
There is nothing wrong with ISL package on conda-forge. GCC 4.8, which is in the default anaconda channel, supports only ISL 0.12.2. GCC 5, for example, supports 0.13 .. 0.14, and GCC 6 - 0.14 .. 0.16. It is just a poor GCC package on anaconda, which doesn't pin supported versions of ISL at all, as well as cloog, mpc, and mpfr.

**Comment by ocefpaf** [comment]
@frol the package might be OK but the "stack" is not. Since we don't package `gcc`, and this `isl` version is picked up by conda, we do have an issue that needs resolving!

Maybe it is just a matter of pinning `isl` where `gcc` is needed, but that might break stacks installing software that uses conda's `gcc` along side software that needs the `conda-forge`'s `isl`.

Let's keep this issue open and discuss possible solutions here.

**Comment by frol** [comment]
@ocefpaf The issue you brought is irrelevant to the original issue. ISL is not faulty, at least as far as I can tell. Also, it is by design that you cannot install two versions of one library simuntaneously in conda. It is not ISL library problem.

**Comment by jakevdp** [comment]
Is there any way in conda to specify the opposite of dependencies? i.e. combinations of library versions that we know will fail together, without forcing installation of a certain version?

**Comment by ocefpaf** [comment]
> ISL is not faulty, at least as far as I can tell.

I renamed the issue to keep the discussion xref'ed to the source. Feel free to close this one and open a new if you think it is important. But please let's discuss this as it is very important for the channel stability.

> Also, it is by design that you cannot install two versions of one library simultaneously in conda. It is not ISL library problem.

Indeed. That is why I want to discuss this. We need to find a solution for packages that need `gcc` and will end up installing this `isl`.

**Comment by ocefpaf** [comment]
> Is there any way in conda to specify the opposite of dependencies? i.e. combinations of library versions that we know will fail together, without forcing installation of a certain version?

Pining will do.

The problem is that you won't ensure that all the software in the stack will be installed together and that is why we need to discuss this, to find a long term solution.

**Comment by frol** [comment]
There are only two package managers out there that I'm aware of solving this kind of problem:
- [Nix](https://nixos.org/nix/) sets up every package independently and injects dependencies with symlinks.
- [NPM](https://www.npmjs.com/) uses nested folders structure if dependencies are in conflict.

**Comment by ocefpaf** [comment]
Thanks for bringing those up @frol.  Bare in mind that I am not asking for a fix here or for this package to be removed (or any other action beyond discussion).

Since we don't provide a `gcc` package let's keep this open so we can point people to the current workaround: pin ISL (`isl 0.12.*`) to ensure the default channel `gcc` works :wink: 

PS: We do need to worry about overall consistency in `conda-forge` and `defaults` and some proposal are already emerging (see https://github.com/conda-forge/libpng-feedstock/issues/7).

---

Edit: open issues help us remember things. This is basically the same as #1, but that one was closed and I did not notice it when navigating there looking for answers.

**Comment by jakirkham** [comment]
Just to help clarify this issue, the problem is the `gcc` package has bad pinnings. It did not pin `isl` at all, which is why this issue emerged. In fact, @frol had added a fix in PR ( https://github.com/conda/conda-recipes/pull/501 ) that was accepted, which would fix this issue. However, the `gcc` package has not been rebuilt since. So, the current `gcc` package in `defaults` still has this problem.

To try to workaround this issue, I added `cloog` with a higher build number and pinned `isl` correctly. As `gcc` 4.8.x depends on `cloog` and `cloog` does depend on `isl`, this didn't introduce any new dependencies, allowed us to build this package too, and fix the pinning as intended. Given that `cloog` was already pinned to a specific version exactly, it would causes us no problems to pin the build number if needed. Pinning the build number was our initial fix, but it turned out we were able to avoid this partially based on how the solver handled duplicates between channels. This seemed like a weak fix, but has held for sometime.

In order to mitigate future damage, we migrated everyone we could off of the `gcc` package. As we had decided to use devtoolset-2.1 at the time as it was the most acceptable solution to everyone involved, this solved the bulk of the issues. C, C++ (even with C++11 support) could be built with that compiler. However, we wanted to still be able to ship `libgomp` and `libgfortran`. So, we continued to use the `gcc` package for those cases.

It appears that there are some changes in the solver that have caused our `cloog` workaround to fail. So, we have a few options. The fallback we had decided on previously was to pin `cloog` exactly down to the build number. As it is already pinned to `cloog` version 0.18.0 exactly, it is not a problem for us to pin the build number too. This ends up being the most flexible solution.

In short, the fix is to add the following to build requirements when using `gcc`.

```
    - cloog 0.18.0 10
```

**Comment by jakirkham** [comment]
Would it be reasonable to have `gcc`'s requirements hot-fixed, @mcg1969? The unpinned `isl` seems to be causing people issues again. Seems like this would be the easiest path forward to fixing that.

**Comment by mingwandroid** [comment]
@jakirkham What do you mean by hot-fixed?

In cases like this, where `gcc` is quite important to have and tied to old versions of packages and therefore holding back cutting-edge package versions, I think we should consider creating legacy versions of the packages what only contain static libraries, so they can be listed as build dependencies only and not as run dependencies.

Of course static linking sometimes introduces licensing issues, but not in this case. If we can't statically link then we should also consider renaming to `isl-legacy` and putting it in a different prefix.

I appreciate It's horrible but I can't see any workable alternatives (shy of updating all the things to very recent versions). I had to do this with `nloptr` recently for R, in this case partly because I didn't want R to pull in Python and partly because `nloptr` on `conda-forge` will probably be built with MSVC and a more recent GCC on the other systems than the rest of the R ecosystem currently uses.

**Comment by mcg1969** [comment]
@mingwandroid, what we mean by a "hotfix" is to alter the `repodata.json` entry for this package so that it fixes its dependencies. @jakirkham, please consider filing an issue in the "anaconda-issues" GitHub issue tracker for this. I've let the team know but that's the right venue for filing and tracking such a request. Of course, refer to this issue when you do.

**Comment by jakirkham** [comment]
Sure, done in this issue ( https://github.com/ContinuumIO/anaconda-issues/issues/833 ), @mcg1969. I'll keep that in mind for future issues of this sort.

**Comment by jakirkham** [comment]
The `gcc` package has been [pinned](https://github.com/ContinuumIO/anaconda-issues/issues/833#issuecomment-228581113) upstream and appears to work. Going to go ahead and close this out. If there are more issues with `gcc`, I would politely recommend that they be raised on the Anaconda [issue tracker](https://github.com/ContinuumIO/anaconda-issues/issues) in the future. Thanks everyone for your patience.

**Comment by jakirkham** [comment]
Side note: Not really sure why this was added, but I notice that many people have `isl` now as a run dependency. This should not have been needed with or without the workaround. `isl` is a dependency of `gcc`, but after a package is built with `gcc` the package is not linked to `isl`. So, there should be no need to distribute `isl` alongside one's package. I would strongly urge people to remove that as a run requirement.
